File: Sorting/Heap_sort.java
import java.util.Arrays;
class Heapify{
public void heapify(int arr[],int i,int modifiedLast){
	if((2*i+1)>modifiedLast)
	  return ;
    if((2*i+2)<modifiedLast && arr[i]<arr[2*i+2])
      (new Build_max_heap()).swap(arr,2*i+2,i);
     if(arr[i]<arr[2*i+1])
     (new Build_max_heap()).swap(arr,2*i+1,i);
}
}
class Build_max_heap{
	 public void buildHeap(int arr[],int i){
		        if ((i-1)/2<0)
		         return;
		         int child=i;
		         int parent=(i-1)/2;
		         if(arr[child]>arr[parent])
		           swap(arr,child,parent);
    }
	public void swap(int arr[],int child,int parent){
	        int temp=arr[child];
	        arr[child]=arr[parent];
	        arr[parent]=temp;
	    }
}
public class Heap_sort
{
   public static void main(String[] args) {
		int arr[]={56,23,67,32,98,76,0};


		Build_max_heap obj=new Build_max_heap();
		for(int j=0;j<arr.length;j=j+1){
		    obj.buildHeap(arr,j);//this function create min heap in O(n) time which is most efficient way to build a heap
	}


	     Heapify hp=new Heapify();
         for(int modifiedLast=arr.length-1;modifiedLast>=0;modifiedLast=modifiedLast-1){
			hp.heapify(arr,0,modifiedLast);
			(new Build_max_heap()).swap(arr,0,modifiedLast);
		}
		System.out.println(Arrays.toString(arr));
		//output:[0,23, 32, 56, 67, 76, 98]
	}
}

File: Sorting/heap-sort.c
#include<stdio.h>
#include<math.h>
int arr[100];
void heapify(int a)
{
    int b=log(a)/log(2);
    printf("a is %d\n",a);
	printf("b is %d\n",b);
    int i,par,ch,temp,ch1,ch2,mxc;
    for(i=pow(2,b)-2;i>-1;i--)
    {
        par=i;    
		printf("parent is   %d\n",i);    
        while(par<=a-1)
        {
            ch1=2*par+1;
            ch2=2*par+2;
            mxc=ch1;
            if(ch1>=a)
                arr[ch1]=-999;
            if(ch2>=a)
                arr[ch2]=-999;
            
            if(arr[ch2]>arr[mxc])
                mxc=ch2;
            
            if(arr[mxc]<arr[par])
                break;
            else
            {
                //swap
                temp=arr[mxc];
                arr[mxc]=arr[par];
                arr[par]=temp;
            }
            par=mxc;
        }                    
    }
}
int main()
{
    int a,b,c,d,e,f;
    scanf("%d",&a);//number of elements in the array
    for(b=0;b<100;b++)
    {
        arr[b]=-100;
    }
    for(b=0;b<a;b++)
    {
        scanf("%d",&arr[b]);
    }arr[a]=-1;
   while(a!=0)
   {
        heapify(a);
        printf("%d  ",arr[0]);
        arr[0]=arr[a-1];
        a--;
   }
   
}
